core/src/main/java/io/questdb/griffin/engine/groupby/DistinctTimeSeriesRecordCursorFactory.java

```diff
@@ -0,0 +1,209 @@
+        RecordSink recordSink = RecordSinkFactory.getInstance(asm, metadata, columnFilter, false);
+        this.dataMap = MapFactory.createMap(configuration, metadata);
+        this.base = base;
+        this.metadata = metadata;
+        this.cursor = new DistinctTimeSeriesRecordCursor(
+                getMetadata().getTimestampIndex(),
+                dataMap,
+                recordSink
+        );
+    }
+
+    @Override
+    public void close() {
+        dataMap.close();
+        base.close();
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) {
+        return cursor.of(base.getCursor(executionContext));
+    }
+
+    @Override
+    public RecordMetadata getMetadata() {
+        return metadata;
+    }
+
+    @Override
+    public boolean recordCursorSupportsRandomAccess() {
+        return base.recordCursorSupportsRandomAccess();
+    }
+
+    private static class DistinctTimeSeriesRecordCursor implements RecordCursor {
+        private final Map dataMap;
+        private final RecordSink recordSink;
+        private final int timestampIndex;
+        private RecordCursor baseCursor;
+        private Record record;
+        private Record recordB;
+        private long prevTimestamp;
+        private long prevRowId;
+        private byte state = 0;
+
+        public DistinctTimeSeriesRecordCursor(int timestampIndex, Map dataMap, RecordSink recordSink) {
+            this.timestampIndex = timestampIndex;
+            this.dataMap = dataMap;
+            this.recordSink = recordSink;
+        }
+
+        @Override
+        public void close() {
+            Misc.free(baseCursor);
+            dataMap.clear();
+        }
+
+        @Override
+        public Record getRecord() {
+            return record;
+        }
+
+        @Override
+        public SymbolTable getSymbolTable(int columnIndex) {
+            return baseCursor.getSymbolTable(columnIndex);
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (state == COMPUTE_NEXT) {
+                while (baseCursor.hasNext()) {
+                    final long timestamp = record.getTimestamp(timestampIndex);
+                    if (timestamp != prevTimestamp) {
+                        prevTimestamp = timestamp;
+                        prevRowId = record.getRowId();
+                        return true;
+                    }
+
+                    if (checkIfNotDupe()) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+            boolean next = state == REUSE_CURRENT;
+            state = COMPUTE_NEXT;
+            return next;
+        }
+
+        @Override
+        public Record getRecordB() {
+            return baseCursor.getRecordB();
+        }
+
+        @Override
+        public void recordAt(Record record, long atRowId) {
+            baseCursor.recordAt(record, atRowId);
```

Comment:
I don't think this is correct, records from this cursor don't map 1 to 1 to base cursor

Reply:
this is tested to be correct :)